### PR TITLE
Added outlier detection with iterativetrimming

### PR DESCRIPTION
**Pull request by swagataroy123**
Hi @jduerholt ,
I have added the robust GP outlier as in this paper [(https://arxiv.org/abs/2011.11057]((https://arxiv.org/abs/2011.11057))

### CODE REVIEW

File: bofire/data_models/outlier_detection/outlier_detection.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+from bofire.data_models.base import BaseModel
+from bofire.data_models.surrogates.api import SingleTaskGPSurrogate
+
+
+class OutlierDetection(BaseModel):
+    type: str
+
+
+class IterativeTrimming(OutlierDetection):
```
Comment (jduerholt) [inline bofire/data_models/outlier_detection/outlier_detection.py:14]: Please add a docstring in which you mention the method and the paper and explain the parameters.

File: bofire/outlier_detection/mapper.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+from typing import Dict, Type
+
+from bofire.data_models.outlier_detection import api as data_models
+from bofire.outlier_detection.outlier_detection import (
+    IterativeTrimming,
+    OutlierDetection,
+)
+
+SURROGATE_MAP: Dict[Type[data_models.OutlierDetection], Type[OutlierDetection]] = {
```
Comment (jduerholt) [inline bofire/outlier_detection/mapper.py:9]:
```suggestion
OUTLIER_MAP: Dict[Type[data_models.OutlierDetection], Type[OutlierDetection]] = {
```

File: bofire/outlier_detection/mapper.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    data_models.IterativeTrimming: IterativeTrimming,
+}
+
+
+def map(data_model: data_models.OutlierDetection) -> OutlierDetection:
+    cls = SURROGATE_MAP[data_model.__class__]
```
Comment (jduerholt) [inline bofire/outlier_detection/mapper.py:15]:
```suggestion
    cls = OUTLIER_MAP[data_model.__class__]
```

File: bofire/outlier_detection/api.py
Comment (jduerholt) [inline bofire/outlier_detection/api.py:1]: add here the import of the outlier detection modeule

File: bofire/outlier_detection/outlier_detection.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    def __init__(self, data_model, **kwargs):
+        self.alpha1 = data_model.alpha1
+        self.alpha2 = data_model.alpha2
+        self.nsh = data_model.nsh
+        self.ncc = data_model.ncc
+        self.nrw = data_model.nrw
+        self.base_gp = data_model.base_gp
+        super().__init__()
+
+    def detect(self, experiments: pd.DataFrame):
```
Comment (jduerholt) [inline bofire/outlier_detection/outlier_detection.py:27]:
```suggestion
    def detect(self, experiments: pd.DataFrame) -> Tuple[pd.DataFrame, pd.DataFrame]:
```

File: tests/bofire/outlier_validation/test_outlier_detection.py
Comment (jduerholt) [inline tests/bofire/outlier_validation/test_outlier_detection.py:1]: call it in the tests exactly as in the the main structure, `outlier_validation` --> `outlier_detection`

File: tests/bofire/outlier_validation/test_outlier_detection.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    ITGP_model = IterativeTrimming(
+        base_gp=SingleTaskGPSurrogate(
+            inputs=inputs, outputs=outputs, kernel=kernel, scaler=scaler
+        )
+    )
+    ITGP = mapper.map(ITGP_model)
+    assert isinstance(ITGP.base_gp, SingleTaskGPSurrogate)
+    assert isinstance(ITGP, mapper.IterativeTrimming)
+    # detect
+    Trimmed, outliers = ITGP.detect(experiments=experiments)
```
Comment (jduerholt) [inline tests/bofire/outlier_validation/test_outlier_detection.py:66]:
```suggestion
    trimmed, outliers = ITGP.detect(experiments=experiments)
```

File: tests/bofire/outlier_validation/test_outlier_detection.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    assert isinstance(Trimmed, pd.DataFrame)
+    assert isinstance(outliers, pd.DataFrame)
+    assert len(experiments) == len(Trimmed) + len(outliers)
```
Comment (jduerholt) [inline tests/bofire/outlier_validation/test_outlier_detection.py:69]: can you also check it it was able to finde most of the outliers?

File: tests/bofire/outlier_validation/test_outlier_detection.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    neal_func(x_ob[:n_outlier]) + np.random.randn(n_outlier) * noise_outlier
+)
+
+experiments = pd.DataFrame()
+experiments["x_1"] = x_ob
+experiments["y"] = y_ob
+experiments["valid_y"] = 1
+
+
+@pytest.mark.parametrize(
```
Comment (jduerholt) [inline tests/bofire/outlier_validation/test_outlier_detection.py:40]: we do not need the parameterization here. just write everything starting in line 20 into  the test_IterativeTrimming method, and remove the argument `experiments` from it.

**Review by jduerholt — CHANGES_REQUESTED** [review]
Thx, I let some comments.

File: bofire/data_models/outlier_detection/api.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+from typing import Union
+try:
+    from bofire.data_models.outlier_detection.outlier_detection import OutlierDetection,IterativeTrimming
+    AnyOutlierDetection = Union[OutlierDetection,IterativeTrimming]
```
[the remark this answers is not in the document]
Comment (swagataroy123) [reply bofire/data_models/outlier_detection/api.py]: I just copied it from api of data_models/surrogates

File: bofire/outlier_detection/api.py
Comment (swagataroy123) [reply bofire/outlier_detection/api.py:1]: yeah ruff messed up something

**Comment by swagataroy123** [comment]
> Thx, I let some comments.

I have answered them

File: bofire/data_models/outlier_detection/outlier_detection.py (hunk 1 above)
Comment (jduerholt) [reply bofire/data_models/outlier_detection/outlier_detection.py:14]: Please use google type docstrings

File: bofire/data_models/outlier_detection/api.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jduerholt) [reply bofire/data_models/outlier_detection/api.py]: It is not needed here, please remove it.

**Comment by jduerholt** [comment]
Please also add a specs module as shown here: https://github.com/experimental-design/bofire/blob/main/tests/bofire/data_models/specs/surrogates.py and add the serialization and deserialization tests.

**Comment by swagataroy123** [comment]
> Furthermore, a notebook under tutorials would be nice in which the ITGP outlier detection stuff is demonstrated based on the ITGP paper.
> 
> Call it outlier_detection.ipynb

**Review by jduerholt — APPROVED** [review]
Looks good to me.

**Pull request merged**